tokio/src/runtime/task/harness.rs

```diff
@@ -49,63 +56,39 @@ where
     pub(super) fn poll(self) {
-        // If this is the first time the task is polled, the task will be bound
-        // to the scheduler, in which case the task ref count must be
-        // incremented.
-        let is_not_bound = !self.core().scheduler.is_bound();
-
-        // Transition the task to the running state.
-        //
-        // A failure to transition here indicates the task has been cancelled
-        // while in the run queue pending execution.
-        let snapshot = match self.header().state.transition_to_running(is_not_bound) {
-            Ok(snapshot) => snapshot,
-            Err(_) => {
-                // The task was shutdown while in the run queue. At this point,
-                // we just hold a ref counted reference. Drop it here.
-                self.drop_reference();
-                return;
-            }
-        };
-
-        if is_not_bound {
-            // Ensure the task is bound to a scheduler instance. Since this is
-            // the first time polling the task, a scheduler instance is pulled
-            // from the local context and assigned to the task.
-            //
-            // The scheduler maintains ownership of the task and responds to
-            // `wake` calls.
-            //
-            // The task reference count has been incremented.
-            //
-            // Safety: Since we have unique access to the task so that we can
-            // safely call `bind_scheduler`.
-            self.core().scheduler.bind_scheduler(self.to_task());
-        }
-
-        // The transition to `Running` done above ensures that a lock on the
-        // future has been obtained. This also ensures the `*mut T` pointer
-        // contains the future (as opposed to the output) and is initialized.
-
-        let waker_ref = waker_ref::<T, S>(self.header());
-        let cx = Context::from_waker(&*waker_ref);
-        let res = poll_future(self.header(), &self.core().stage, snapshot, cx);
-
-        match res {
+        match self.poll_inner() {
             PollFuture::Notified => {
                 // Signal yield
                 self.core().scheduler.yield_now(Notified(self.to_task()));
                 // The ref-count was incremented as part of
                 // `transition_to_idle`.
                 self.drop_reference();
             }
-            PollFuture::Complete(out) => {
-                self.complete(out, snapshot.is_join_interested());
+            PollFuture::DropReference => {
+                self.drop_reference();
+            }
+            PollFuture::Complete(out, is_join_interested) => {
+                self.complete(out, is_join_interested);
             }
             PollFuture::None => (),
         }
     }
 
+    fn poll_inner(&self) -> PollFuture<T::Output> {
+        let snapshot = match self.scheduler_view().transition_to_running() {
+            Ok(snapshot) => snapshot,
+            Err(()) => return PollFuture::DropReference,
+        };
```

Comment:
But when we call it here, we match on `Err(())`, returning `PollFuture::DropReference`.